core/shared/src/main/scala/scalaz/zio/IOQueue.scala

```diff
@@ -38,44 +34,73 @@ class IOQueue[A] private (capacity: Int, ref: IORef[State[A]]) {
    * the queue.
    */
   final def offer[E](a: A): IO[E, Unit] =
-    IO.flatten(ref.modifyFold[E, IO[E, Unit]] {
-      case Deficit(takers) =>
-        takers.dequeueOption match {
-          case None                  => (IO.unit[E], Surplus(Queue.empty[A].enqueue(a), Queue.empty))
-          case Some((taker, takers)) => (taker.complete[E](a).toUnit, Deficit(takers))
-        }
-      case Surplus(values, putters) =>
-        if (values.length < capacity && putters.isEmpty) {
-          (IO.unit[E], Surplus(values.enqueue(a), putters))
-        } else {
```

Comment:
Is there a reason to create an empty queue and enqueue instead of using `Queue.apply[A]`?